agent/agent.go

```diff
@@ -0,0 +1,206 @@
+//  Copyright (c) 2018 Cisco and/or its affiliates.
+//
+//  Licensed under the Apache License, Version 2.0 (the "License");
+//  you may not use this file except in compliance with the License.
+//  You may obtain a copy of the License at:
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+//  Unless required by applicable law or agreed to in writing, software
+//  distributed under the License is distributed on an "AS IS" BASIS,
+//  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+//  See the License for the specific language governing permissions and
+//  limitations under the License.
+
+package agent
+
+import (
+	"os"
+	"os/signal"
+	"syscall"
+
+	"errors"
+
+	"github.com/ligato/cn-infra/core"
+	"github.com/ligato/cn-infra/logging/logrus"
+	"github.com/ligato/cn-infra/utils/once"
+	"github.com/namsral/flag"
+)
+
+// Agent implements startup & shutdown procedures.
+type Agent interface {
+	Run() error
+	Start() error
+	Wait() error
+	After() <-chan struct{}
+	Error() error
+	Stop() error
+	Options() Options
+}
+
+// NewAgent creates a new Agent
+func NewAgent(opts ...Option) Agent {
+	options := newOptions(opts...)
+
+	flag.Parse()
+
+	return &agent{
+		opts: options,
+	}
+}
+
+type agent struct {
+	opts      Options
+	stopCh    chan struct{}
+	startOnce once.ReturnError
+	stopOnce  once.ReturnError
+}
+
+// Options returns the Options the agent was created with
+func (a *agent) Options() Options {
+	return a.opts
+}
+
+// Start starts the agent.  Start will return as soon as the Agent is ready.  The Agent continues
+// running after Start returns.
+func (a *agent) Start() error {
+	return a.startOnce.Do(a.startSignalWrapper)
+}
+
+func (a *agent) startSignalWrapper() error {
+	// If we want to properly handle cleanup when a SIG comes in *during*
+	// agent startup (ie, clean up after its finished) we need to register
+	// for the signal before we start() the agent
+	sig := make(chan os.Signal, 1)
+	signal.Notify(sig, syscall.SIGINT, syscall.SIGTERM, syscall.SIGKILL)
+	err := a.start()
+	// If the agent started, we have things to clean up if here is a SIG
+	// So fire off a goroutine to do that
+	if err == nil {
+		go func() {
```

Comment:
Seems like this function should be extracted instead of `start`.

Reply:
start containers the actual logic for starting things.  Its thematically self contained.  This function thematically contains the logic for signal handling.    Putting both things in the same function becomes much harder to understand for subsequent readers.